src/test/java/lambda/part3/exercise/Exercise4.java

```diff
@@ -0,0 +1,123 @@
+package lambda.part3.exercise;
+
+import lambda.data.Employee;
+import lambda.data.JobHistoryEntry;
+import lambda.data.Person;
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@SuppressWarnings({"unused", "ConstantConditions"})
+class Exercise4 {
+
+    private static class LazyCollectionHelper<T, R> {
+
+        private final List<T> source;
+
+        private final Function<T, List<R>> function;
+
+        public LazyCollectionHelper(List<T> source, Function<T, List<R>> function) {
+            this.source = source;
+            this.function = function;
+        }
+
+        public static <T> LazyCollectionHelper<T, T> from(List<T> list) {
+            return new LazyCollectionHelper<>(list, Collections::singletonList);
+        }
+
+        public <U> LazyCollectionHelper<T, U> flatMap(Function<R, List<U>> flatMapping) {
+            return new LazyCollectionHelper<>(source, this.function.andThen(list -> {
+                List<U> result = new ArrayList<>();
+                list.forEach(flatMapping.andThen(result::addAll)::apply);
+                return result;
+            }));
+        }
+
+        public <U> LazyCollectionHelper<T, U> map(Function<R, U> mapping) {
+            return new LazyCollectionHelper<>(source, this.function.andThen(list -> {
+                List<U> result = new ArrayList<>();
+                list.forEach(mapping.andThen(Collections::singletonList).andThen(result::addAll)::apply);
```

Comment:
Do we need `andThen(Collections::singletonList)` here? It is not complaining, I am just interested because this task is quit hard to understand)